mypy/typeshed/stdlib/builtins.pyi

```diff
@@ -40,6 +39,7 @@ from typing import (  # noqa: Y022
     _NamedCallable,
     IO,
     Any,
+    AnyStr,
```

Comment:
Why not `TypeVar("_PathT_co", bound=str | bytes, covariant=True)`?

Reply:
uhh, i don't really know anything about it, but i think `PathLike` is either `str` or `bytes`, never both?

Reply:
True, but now there's no way to type `PathLike[str | bytes]` if you don't care about the input, forcing you to use `Any`.

Reply:
A disjoint union is what would've been best here: https://en.wikipedia.org/wiki/Tagged_union

Reply:
> True, but now there's no way to type `PathLike[str | bytes]` if you don't care about the input, forcing you to use `Any`.

`PathLike[str] | PathLike[bytes]`

Reply:
> but now there's no way to type `PathLike[str | bytes]` 

not sure i follow, that's invalid and is in mypy too
https://mypy-play.net/?mypy=latest&python=3.12&gist=bcd7f8d13cb6c1678e3151f3900d1b46

also:

```py
PathLike[str] | PathLike[bytes]
```

https://github.com/KotlinIsland/basedmypy/pull/817#discussion_r1862851817

Reply:
> not sure i follow, that's invalid and is in mypy too https://mypy-play.net/?mypy=latest&python=3.12&gist=bcd7f8d13cb6c1678e3151f3900d1b46

Well, yea...
My point is that I don't think it should be illegal at all. Cus even though `str & bytes` isn't literally allowed, `str | bytes` is the best way we can currently express something that is "either `str` or `bytes`". It would be even better if there'd be a way to express `str ^ bytes` with which I mean `str | bytes - str & bytes`. But since we don't have that, I think we should use the next best *valid* alternative that we have at hand. And as I briefly explained earlier, I don't think that `T: (str, bytes)` is valid as a generic type parameter, as it is neither *co-*, *contra-*, nor *in*variant. Or on other words; it doesn't *vary* with the input; it *selects* that what is closest to it. So while that's valid when used as syntactic sugar for overloads, it shouldn't be confused with a type parameter. Because by doing so, you're dismissing the most fundamental and essential rule of type systems: Types don't change. But if you allow it anyway, and declare the type `PathLike[SubtypeOfStr]`, then it immediately changes into `PathLike[str]` -- Haram!

Reply:
> ```python
> PathLike[str] | PathLike[bytes]
> ```

That means that instead of

```pyi
def to_path[T: str | bytes](x: T, /) -> ToPath[T]: ...
```

you're must write it as

```pyi
@overload
def to_path(x: str, /) -> PathLike[str]: ...
@overload
def to_path(x: bytes, /) -> PathLike[bytes]: ...
@overload
def to_path(x: str | bytes, /) -> PathLike[str] | PathLike[bytes]: ...
```

So that defeats the purpose of the single use-case for which the "type-var with constraints" is intended in the first place; compressing overloads. Actually, it's even worse than that; it achieves the exact opposite.

Reply:
```py
def to_path[T: (str, bytes)](x: T, /) -> ToPath[T]: ...
```

Reply:
that doesn't include the `x: str | bytes` case

Reply:
hmmm, I suppose a single value of `str | bytes` will always be `str` or `bytes`, so that's an improvement that could be made, but:

```py
def f[T: str|bytes](T1: T, T2: T) -> PathLike[T]:...
```
this would be really unsafe

Reply:
> this would be really unsafe

so `class PathLike[T: str | bytes](Protocol): ...` ?

Reply:
```py
def f[T: (str, bytes)](t: T) -> PathLike[T]:...

d: str | bytes
r = f(d)
reveal_type(r) # PathLike[str] | PathLike[bytes]
```
here we can see that the union distributes over the function and we get the desired behaviour

Reply:
> ```python
> def f[T: (str, bytes)](t: T) -> PathLike[T]:...
> 
> d: str | bytes
> r = f(d)
> reveal_type(r) # PathLike[str] | PathLike[bytes]
> ```
> 
> here we can see that the union distributes over the function and we get the desired behaviour

That doesn't work it pyright though 🤔

Reply:
mmm, it's a limitation. I have an issue somewhere to do unions distribution properly, I think this might be a simple case that can be solved in a follow up

Reply:
> * [union distrubution of calls doesn't apply to constraints #831](https://github.com/KotlinIsland/basedmypy/issues/831)

As I explained in the issue, that's not valid, because it doesn't take the intersection type into account.